src/ray/raylet/scheduling/local_resource_manager.h

```diff
@@ -0,0 +1,218 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <gtest/gtest_prod.h>
+
+#include <iostream>
+#include <sstream>
+#include <vector>
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/container/flat_hash_set.h"
+#include "ray/common/task/scheduling_resources.h"
+#include "ray/gcs/gcs_client/accessor.h"
+#include "ray/gcs/gcs_client/gcs_client.h"
+#include "ray/raylet/scheduling/cluster_resource_data.h"
+#include "ray/raylet/scheduling/cluster_resource_scheduler_interface.h"
+#include "ray/raylet/scheduling/fixed_point.h"
+#include "ray/raylet/scheduling/scheduling_ids.h"
+#include "ray/raylet/scheduling/scheduling_policy.h"
+#include "ray/util/logging.h"
+#include "src/ray/protobuf/gcs.pb.h"
+
+namespace ray {
+
+using rpc::HeartbeatTableData;
+
+/// Class manages the resources of the local node.
+class LocalResourceManager {
+ public:
+  /// Constructor initializing the resources associated with the local node.
+  ///
+  /// \param local_node_id: ID of local node,
+  /// \param local_node_resources: The total and the available resources associated
+  /// with the local node.
+  LocalResourceManager(int64_t local_node_id, StringIdMap &resource_name_to_id,
+                       const NodeResources &local_node_resources);
+  LocalResourceManager(
+      int64_t local_node_id, StringIdMap &resource_name_to_id,
+      const absl::flat_hash_map<std::string, double> &local_node_resources,
+      std::function<int64_t(void)> get_used_object_store_memory = nullptr,
+      std::function<bool(void)> get_pull_manager_at_capacity = nullptr);
+
+  /// Add a local resource that is available.
+  ///
+  /// \param resource_name: Resource which we want to update.
+  /// \param resource_total: New capacity of the resource.
+  void AddLocalResourceInstances(const std::string &resource_name,
```

Comment:
Add & Create are different?

Reply:
Great question... i don't know

Reply:
Here `Add` means raylet receives an `add` request and then add the resource name and capacity into map. `Create` more like means raylet generate them by itself. Just personal idea.